app/services/allocation/AssessorAllocationService.scala

```diff
@@ -16,67 +16,133 @@
 
-  def allocationRepo: AssessorAllocationMongoRepository
-  val eventsService: EventsService
+  val authProviderClient = AuthProviderClient
+  val emailClient = CSREmailClient
+}
 
+trait AssessorAllocationService extends EventSink {
 
+  def assessorAllocationRepo: AssessorAllocationMongoRepository
   def candidateAllocationRepo: CandidateAllocationMongoRepository
+  val applicationRepo: GeneralApplicationRepository
+
+  def eventsService: EventsService
+
+  def emailClient: EmailClient
+  def authProviderClient: AuthProviderClient
 
   def getAllocations(eventId: String): Future[exchange.AssessorAllocations] = {
-    allocationRepo.allocationsForEvent(eventId).map { a => exchange.AssessorAllocations.apply(a) }
+    assessorAllocationRepo.allocationsForEvent(eventId).map { a => exchange.AssessorAllocations.apply(a) }
   }
 
   def getCandidateAllocations(eventId: String): Future[exchange.CandidateAllocations] = {
     candidateAllocationRepo.allocationsForEvent(eventId).map { a => exchange.CandidateAllocations.apply(a) }
   }
 
   def allocate(newAllocations: command.AssessorAllocations): Future[Unit] = {
-    allocationRepo.allocationsForEvent(newAllocations.eventId).flatMap {
-      case Nil => allocationRepo.save(persisted.AssessorAllocation.fromCommand(newAllocations)).map(_ => ())
+    assessorAllocationRepo.allocationsForEvent(newAllocations.eventId).flatMap {
+      case Nil => assessorAllocationRepo.save(persisted.AssessorAllocation.fromCommand(newAllocations)).map(_ => ())
       case existingAllocations => updateExistingAllocations(existingAllocations, newAllocations).map(_ => ())
     }
   }
 
-  def allocateCandidates(newAllocations: command.CandidateAllocations): Future[Unit] = {
-    getCandidateAllocations(newAllocations.eventId).flatMap { existingAllocation =>
-      existingAllocation.allocations match {
-        case Nil => candidateAllocationRepo.save(persisted.CandidateAllocation.fromCommand(newAllocations)).map(_ => ())
-        case _ => updateExistingAllocations(existingAllocation, newAllocations).map(_ => ())
+  private val dateFormat = "dd MMMM YYYY"
+  private val timeFormat = "HH:mma"
+
+
+  def allocateCandidates(newAllocations: command.CandidateAllocations)(implicit hc: HeaderCarrier, rh: RequestHeader): Future[Unit] = {
+
+    eventsService.getEvent(newAllocations.eventId).flatMap { event =>
+      val eventDate = event.date.toString(dateFormat)
+      val eventTime = event.startTime.toString(timeFormat)
+      val deadlineDateTime = event.date.minusDays(10).toString(dateFormat)
+
+
+      getCandidateAllocations(newAllocations.eventId).flatMap { existingAllocation =>
+        existingAllocation.allocations match {
+          case Nil =>
+            Future.sequence(newAllocations.allocations.map(sendCandidateEmail(_, eventDate, eventTime, deadlineDateTime)))
+              .flatMap { _ =>
+                candidateAllocationRepo.save(persisted.CandidateAllocation.fromCommand(newAllocations)).map(_ => ())
```

Comment:
I think the order here is wrong - we should save the allocations before sending the email, otherwise if things go wrong we may invite a candidate without having an allocation for them - could be embarrassing.